### ISSUE TITLE
Add support for compressed tarfiles

### ISSUE THREAD

**Issue by jose1711** [mxmlnkn/ratarmount#7]
Not sure if it's planned but I would very much welcome it. Performance of `archivemount` is disappointing.

**Comment by mxmlnkn** [comment]
I would welcome it and it seems to be possible in general, see https://github.com/tomorrow-nf/tar-as-filesystem/ but in the short-term I don't think, I will be able to implement it. Merge requests are welcome and I'll leave this issue open.

**Comment by al-zatv** [comment]
may be pixz (or may be ideas from pixz) will help 
https://github.com/vasi/pixz

**Comment by mxmlnkn** [comment]
@al-zatv Thanks. I also have [seek-bzip2](https://bitbucket.org/james_taylor/seek-bzip2), [lzopfs](https://github.com/mxmlnkn/lzopfs), and even the possibly insane idea to abuse the python3 internal [bzip2 module](https://github.com/python/cpython/blob/530f506ac91338b55cf2be71b1cdf50cb077512f/Modules/_bz2module.c) in some way. As far as I understand bzip2 is just a simple "array" of fixed-length blocks. I'd only need to cache which blocks corresponds to which uncompressed offsets, create a lookup table, and then seek to the correct block.

However, one major problem I currently see, is that most of these a C/C++ based solutions, which I definitely don't wanna reimplement in Python. However, it's kinda nice that currently, the program is basically just one self-contained Python script. Also, I don't have any experience packaging C/Python mixed applications, yet. It might also be better to just rewrite the python part in C++ and go full C++. These are the things, I'll have to ponder and which might hinder my motivation to implement it in any way. 

Feedback would be welcome. I hope to some time in November during my vacation.

**Comment by al-zatv** [comment]
In my use cases, I don't create such archives for use with ratarmount. I receive them, and often they are too large to unpack them on my disk, so I ended up using ratarmount.

Users like me will need popular formats mainly. Most popular is tar.gz, I think.

As for modularity, I also think that current situation is very convinient. I just do wget for fresh file, and I'm all set.

**Comment by ole-tange** [comment]
I think this will be doable for compression methods that take a full block and compresses it (e.g bzip2) but near to impossible to do for compression methods that reuse a dictionary while compressing the next byte (e.g. gzip).

If you are using tar.gz from others, all is not lost: You simply recompress them when you fetch them:

```
wget -O - https://...tar.gz | zcat | pbzip2 > my.tar.bz2
```

**Comment by al-zatv** [comment]
> If you are using tar.gz others have made, all is not lost: You simply recompress them when you fetch them:
> 
> ```
> wget -O - https://...tar.gz | zcat | pbzip2 > my.tar.bz2
> ```

Nice idea! May be once I apply it!  
But usually it is not practical for me: files are too big, so I download them by parts (to have chance to re-download separate parts if something goes wrong). 
But, of course, bz2 support may help!

**Comment by mxmlnkn** [comment]
@al-zatv I think, I won't be able to do gzip support even if that is most used in your case.

I'm currently [working](https://github.com/mxmlnkn/ratarmount/commit/540f50c3117adffde317a411f7f2886746ee1f40) on Bzip2 support based on the [0-BSD licensed](https://landley.net/toybox/license.html) version of [micro-bunzip](https://landley.net/code/) as it comes with [toybox](http://landley.net/hg/toybox/file/37ea9dff9c27/lib/bunzip.c). It's quite interesting to see some of the innards of the bzip2 format. Small caveat: the code does not seem to support pbzip2 created archives but maybe I'll be able to fix that at some point. The heart of the code will have to be in C/C++ for performance. But now that #11 has been merged, I feel a bit better about bundling C/C++/Cython code. In the end, it's like a simple python-only script is cross-platform anyways, especially seeing that Windows does not have good FUSE / fusepy support.

> By the way, may be there is fuse module for "gz" or "bz2" or,better,for both of them? May be it will allow to put ratarmount on top of it. Or,may be,its stupid idea, fuse module for single file:)

Hm, I see what you mean. A FUSE module wouldn't be necessary or really possible as there is only one file inside without any metadata. I'm currently writing a python file object module, which supports block seeking as basically what you mean with FUSE just a bit stripped down. I'll have to implement read,write,seek,tell and such things but no FUSE specific things like readdir,getattrinfo, ...

**Comment by mxmlnkn** [comment]
I have a [working version](https://github.com/mxmlnkn/ratarmount/pull/15) now. If anyone wants to test it, feedback is welcome, especially as it now comes with a C++ source. Installation steps are still the same:
```bash
git clone https://github.com/mxmlnkn/ratarmount.git
git checkout develop
python3 -m pip install --user .

# Tests
wget 'https://ftp.mozilla.org/pub/firefox/releases/70.0/linux-x86_64/en-US/firefox-70.0.tar.bz2'
time tar -xf firefox-70.0.tar.bz2 # 10.1s
time ratarmount firefox-70.0.tar.bz2 # 10.3s
time diff --recursive firefox firefox-70.0/firefox/ # 68.4s
time cat firefox/updater.ini # 2ms
time cat firefox-70.0/firefox/updater.ini # 45ms
```

It's not perfect seeking as it only seeks blockwise and then has to decode the block up to the specified seek offset. But, a block is only 100-900kB. So, for anything larger, there will be a speedup proportional to the block count. As an example, the above firefox.tar.bz2 is 65M and has 228 blocks. The block count can be queried with:
```bash
python3 -c '
    import ratarmount
    blockOffsets = ratarmount.SeekableBzip2( "firefox-70.0.tar.bz2" ).blockOffsets()
    print( "BZip2 Blocks: ", len( blockOffsets ) )'
```

**Comment by ole-tange** [comment]
Dumps core on my system:

~~~
$ python3 -m pip install --user .
Processing /mnt/4tb/home/tange/tmp/ratarmount
Collecting cbor (from ratarmount==0.3.1)
Collecting fusepy (from ratarmount==0.3.1)
Collecting indexed_gzip (from ratarmount==0.3.1)
  Using cached https://files.pythonhosted.org/packages/5b/e8/1472b03a6c3db08d46383b81bf1cc955752540370f4e3879acfe102bfac5/indexed_gzip-0.8.10-cp36-cp36m-manylinux1_x86_64.whl
Collecting lz4 (from ratarmount==0.3.1)
  Using cached https://files.pythonhosted.org/packages/5d/5e/cedd32c203ce0303188b0c7ff8388bba3c33e4bf6da21ae789962c4fb2e7/lz4-2.2.1-cp36-cp36m-manylinux1_x86_64.whl
Collecting msgpack (from ratarmount==0.3.1)
  Using cached https://files.pythonhosted.org/packages/3d/a8/e01fea81691749044a7bfd44536483a296d9c0a7ed4ec8810a229435547c/msgpack-0.6.2-cp36-cp36m-manylinux1_x86_64.whl
Collecting python-rapidjson (from ratarmount==0.3.1)
  Using cached https://files.pythonhosted.org/packages/61/2c/ec819d4603da706a80c4b20583b2ed7df89b6a0fc8467f090b45908f905c/python_rapidjson-0.9.1-cp36-cp36m-manylinux1_x86_64.whl
Collecting pyyaml (from ratarmount==0.3.1)
Collecting simplejson (from ratarmount==0.3.1)
Collecting ujson (from ratarmount==0.3.1)
Installing collected packages: cbor, fusepy, indexed-gzip, lz4, msgpack, python-rapidjson, pyyaml, simplejson, ujson, ratarmount
  Running setup.py install for ratarmount ... done
Successfully installed cbor-1.0.0 fusepy-3.0.1 indexed-gzip-0.8.10 lz4-2.2.1 msgpack-0.6.2 python-rapidjson-0.9.1 pyyaml-5.1.2 ratarmount-0.3.1 simplejson-3.17.0 ujson-1.35
Segmentation fault (core dumped)

**Comment by mxmlnkn** [comment]
I don't understand this... why should it segfault **after** having successfully installed all packages including ratarmount.

Can you use it anyways after this botched installation? `ratarmount --version`, ...

Can you try again with the verbose option? `python3 -m pip --verbose install --user .`

Could you do a backtrace with gdb: `gdb --args  python3 -m pip install --user .` and then in the appearing command line press `r` and enter and then after it crashes `bt` and enter?

**Comment by ole-tange** [comment]
See output of all in:

https://cloud.prosa.dk/s/WMceN9iW6eqWnCt

If no one else can reproduce the error, let us assume my system is sick.

**Comment by al-zatv** [comment]
Ubuntu 16.04, failed on install:

  Running setup.py bdist_wheel for ratarmount ... error
  Complete output from command /usr/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-req-build-oh1z5qwx/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/pip-wheel-b89_kcfp --python-tag cp35:
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-3.5
  copying ratarmount.py -> build/lib.linux-x86_64-3.5
  running build_ext
  building 'bzip2' extension
  creating build/temp.linux-x86_64-3.5
  x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I. -I/usr/include/python3.5m -c bzip2.cpp -o build/temp.linux-x86_64-3.5/bzip2.o -std=c++11 -O3 -DNDEBUG
  cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
  In file included from bzip2.cpp:659:0:
  bzip2.h:310:91: error: expected ‘,’ before ‘)’ token
           static_assert( std::numeric_limits<decltype( symbolCount )>::max() >= MAX_SYMBOLS );
                                                                                             ^
  bzip2.h:310:91: error: expected string-literal before ‘)’ token
  error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
  
  ----------------------------------------
  Failed building wheel for ratarmount

**Comment by mxmlnkn** [comment]
@al-zatv 
If that is the only error, then simply adding a message to the `static_assert` or deleting it might already be enough. Here is a version without that assert:
[ratarmount-0.3.2.tar.gz](https://github.com/mxmlnkn/ratarmount/files/3884156/ratarmount-0.3.2.tar.gz)
You should be able to install it directly with pip: `pip3 install --user https://github.com/mxmlnkn/ratarmount/files/3884156/ratarmount-0.3.2.tar.gz`

**Comment by mxmlnkn** [comment]
> See output of all in:
> 
> https://cloud.prosa.dk/s/WMceN9iW6eqWnCt
> 
> If no one else can reproduce the error, let us assume my system is sick.

It looks like ratarmount 0.3.1 was installed correctly as the version output works. It might be that it works correctly, try and test it with any .tar.bz2 or .tar.gz.

The segfault is inside the python garbage collector, so I don't think I have much control over that. Pip shouldn't execute any ratarmount code except for the setup.py and that file is pretty simple. 

The `error: invalid command 'bdist_wheel'` error inside the venv you have [seems to be because you have to install wheels](https://stackoverflow.com/questions/34819221/why-is-python-setup-py-saying-invalid-command-bdist-wheel-on-travis-ci) inside the venv first with `pip install wheel`. 

It might be that this problem does not appear with the other packages because they a wheel package on PyPI. Ratarmount is currently source (.tar.gz) only. Therefore, it has to be built first on any system. I'll have to look into how to create a many-linux-wheel, currently, I only know how to build a x86_64 wheel and PyPI does not accept that.

**Comment by mxmlnkn** [comment]
> may be pixz (or may be ideas from pixz) will help
> https://github.com/vasi/pixz

This is what pixz says:

> The existing XZ Utils provide great compression in the .xz file format, but they produce just one big block of compressed data. Pixz instead produces a collection of smaller blocks which makes random access to the original data possible. This is especially useful for large tarballs.

This makes me believe that xz seeking support for "normal" .tar.xz files isn't possible anyways. I could only support pixz created files but for now that is not planned.

As bzip2 and gzip support are beginning to look like they work after some hiccups like in #17, I'll close this issue for now.

**Comment by rickhg12hs** [comment]
Just wondering ... Curious.

How does a xz compressed squashfs work?  FYI, the mounted squashfs version of the large .tar.bz2 file from #17 can be read about twice as fast as the `ratarmount` .tar.bz2 fuse mount.

**Comment by mxmlnkn** [comment]
> How does a xz compressed squashfs work? 

Squashfs [project site](https://sourceforge.net/projects/squashfs/) summary:

> Squashfs compresses both files, inodes and directories, and supports block sizes up to 1Mbytes for greater compression

It seems to work similarly to pixz but because it is a filesystem, the compressed blocks are not file agnostic. So, seeking to the start of a file should not require seeking inside a compressed block like is required for .tar.bz2 files. You can just jump to the beginning of compressed blocks, which coincides with a file system block. Although, reading the [readmes](https://sourceforge.net/p/squashfs/code/ci/master/tree/RELEASE-READMEs/), it looks like it has become more sophisticated with the introduction of [fragment blocks](https://sourceforge.net/p/squashfs/code/ci/master/tree/RELEASE-READMEs/README-2.0) and such to improve compression for files smaller than the block size

> FYI, the mounted squashfs version of the large .tar.bz2 file from #17 can be read about twice as fast as the `ratarmount` .tar.bz2 fuse mount.

Challenge accepted :laughing:.

What did you measure? Seek time latency with `head -c 1 file` or decoding bandwidth with, e.g., `cat file | wc -c`?

As noted in the new `indexed_bzip2` repo's [issue](https://github.com/mxmlnkn/indexed_bzip2/issues/1), the bz2 decoder is not optimized very well and could be parallelized in some ways. Wishfully, that would get a speedup by the number of cores for sequential reading and some speedup for in-block seeking to reduce the seek latency.

Squashfs seems to have multithreaded compression support since [version 3.1](https://sourceforge.net/p/squashfs/code/ci/master/tree/RELEASE-READMEs/README-3.1#l18).

The seek latency is a bit harder to optimize. After all this script started out more as a proof of concept hacked together personal script and has some high level baggage like the SQL, Python, Cython, and FUSE layers.

But, ratarmount is not intended to compete with squashfs directly. If you can use squashfs, then I'd recommded it (never used it myself, though). But, it's even in the kernel. Ratarmount is more for .tar, .tar.bz2, .tar.gz files you have lying around already.

**Comment by rickhg12hs** [comment]
> Challenge accepted :laughing:.
> What did you measure?

First I mounted the tar.bz2 file using `ratarmount` and then I created the `squashfs` filesystem.

```
$ ratarmount -c --fuse uid=$(id -u),gid=$(id -g),ro CTU-13-Dataset.tar.bz2
$ mksquashfs CTU-13-Dataset CTU-13-Dataset-xz.squashfs -comp xz -info -progress -b 1M -Xdict-size 100% -processors 4
```
FYI, the `xz` `squashfs` is smaller than the .tar.bz2
```
$ stat -c '%s %n' CTU-13-Dataset.tar.bz2 CTU-13-Dataset-xz.squashfs 
1997547391 CTU-13-Dataset.tar.bz2
1846898688 CTU-13-Dataset-xz.squashfs
```
Then I `mount`'ed the `squashfs` filesystem.
```
$ sudo mount ./CTU-13-Dataset-xz.squashfs mnt
```
And then I timed file checksum generation.
```
$ for f in $(find CTU-13-Dataset/ -type f); do /usr/bin/time -f '%e' sha256sum $f; /usr/bin/time -f '%e' sha256sum mnt/${f#*/}; done
```
Sadly, I didn't store the results, but with small files there was no measured read time difference.  The large files showed the greatest time difference and not always with the same time ratio.

It wasn't the best experiment.  All the files were on the same physical medium, so that's good.  However, one compressor was `bzip2` with unknown level (128k blocks?), the other compressor was `xz` with the largest `squashfs` blocksize (`1M`) with an equally sized dictionary.  So, kinda apples and oranges really.  But, if I need to read those pcap files a lot, the `squashfs-xz` is the way to go.  There may be even better compressor algorithm/parameters - I didn't check.

FYI, I also created a `zstd` compressed `squashfs` with a level set somewhere between 6 and 9 inclusive, but that file was slightly larger that the .tar.bz2 file.  Interestingly, the file read speed of the `squashfs-zstd` was faster than the `squashfs-xz`.  Sorry, I don't remember the exact `mksquash` parameters when testing the `zstd` compressor.
